app/services/analysis/gwas.rb

```diff
@@ -0,0 +1,102 @@
+class Analysis
+  class Gwas
+    def initialize(analysis, runner: nil)
+      @analysis = analysis
+      @runner = runner
+    end
+
+    def call
+      prepare_csv_data_files
+      runner.call(job_command) do
+        store_results
+      end
+    end
+
+    private
+
+    attr_reader :analysis
+
+    def prepare_csv_data_files
+      return if genotype_data_file(:csv)
+
+      vcf_data_file = genotype_data_file
+
+      genotype_csv, map_csv = Analysis::Gwas::GenotypeVcfToCsvConverter.new.call(vcf_data_file.file.path)
+
+      analysis.data_files.create!(
+        role: :input,
+        data_type: :gwas_genotype,
+        file: genotype_csv,
+        file_content_type: "text/csv",
+        owner: analysis.owner
+      )
+
+      analysis.data_files.create!(
+        role: :input,
+        data_type: :gwas_map,
+        file: map_csv,
+        file_content_type: "text/csv",
+        owner: analysis.owner
+      )
+    end
+
+    def runner
+      @runner ||= Analysis::ShellRunner.new(analysis)
+    end
+
+    def store_results
+      selected_traits.each do |trait|
+        runner.store_result("SNPAssociation-Full-#{trait}.csv", data_type: :gwas_results)
+      end
+    end
+
+    def selected_traits
+      @selected_traits ||= analysis.args["phenos"]
+
+      unless @selected_traits.present?
+        file = File.open(phenotype_data_file.file.path, "r")
+        result = Analysis::Gwas::PhenotypeCsvParser.new.call(file)
+        @selected_traits = result.trait_ids
+      end
+
+      @selected_traits
+
+    ensure
+      file && file.close
```

Comment:
I see that you use this construct several times. Wouldn't it be as good to use the block version of File.open? It should save some source lines. https://ruby-doc.org/core-2.2.0/File.html#method-c-open